packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -1466,6 +1469,35 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       );
     }
 
+    tsTryParseTypeArguments(): ?N.TsTypeParameterInstantiation {
+      if (!this.isRelational("<")) {
+        return undefined;
+      }
+      const state = this.state.clone();
+      const node = this.startNode();
+      const params = this.tsInType(() =>
+        // Temporarily remove a JSX parsing context, which makes us scan different tokens.
+        this.tsInNoContext(() => {
+          this.next();
+          return this.tsTryParseDelimitedList(
+            "TypeParametersOrArguments",
+            this.tsParseType.bind(this),
+          );
+        }),
+      );
+      if (params) {
+        node.params = params;
+        // This reads the next token after the `>` too, so do this in the enclosing context.
+        // But be sure not to parse a regex in the jsx expression `<C<number> />`, so set exprAllowed = false
+        this.state.exprAllowed = false;
+        if (this.eatRelational(">")) {
+          return this.finishNode(node, "TSTypeParameterInstantiation");
+        }
+      }
+      this.state = state;
+      return undefined;
+    }
+
```

Comment:
This is not nice, as we have this code now twice. But I see why you have done this.

What you should consider here is that we now clone the state twice. Once here and once probably in `tsTryParseDelimitedList`. Cloning the state is also something that is more on the slow side, but not sure how it compares to throwing exceptions.